cpp/onedal/detail/storage_iface.hpp

```diff
@@ -0,0 +1,66 @@
+/*******************************************************************************
+ * Copyright 2020 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+#pragma once
+
+#include "onedal/array.hpp"
+
+#include <utility>
+
+namespace dal::detail {
+
+struct storage_readable {};
+struct storage_writable {};
+struct storage_readable_writable {};
+
+template <typename StorageType>
+class dense_storage_iface {};
+
+template <>
+class dense_storage_iface<storage_readable> : public base {
```

Comment:
Overall, I realised that all interfaces are inherited from `base`. This is not quite right, because `base` is supposed to **contain implementation**. It's not just some abstract principle, but lead to a real problem.

In particular, one of the reason I've introduced `base` is that it should/may overloads new/delete for all its descendants -- this is implementation.

Generally, we discourage multiple public inheritance of implementation classes in DAAL, but allow multiple interface inheritance (say hello to Java/C#), which is, as you now, helps us to escape from the diamond problem. **But**, if multiple (`I1` and `I2`) interfaces are inherited from `base` and required for class `A : I1, I2`, it creates the diamond with `base`, `I1`, `I2` and `A`.

The right `base` should look as follows:
```cpp
class iface {
public:
   virtual ~iface() = default;
   virtual void foo() = 0;
};

class impl : public iface, public base {
public:
   void foo() override;
};
```